### DISCUSSION TITLE
Any idea to manage the constants properly?

### DISCUSSION THREAD

**Discussion by AndrewLawrence80** [golang-standards/project-layout#254] [Q&A]
As programs grow bigger, it is really annoying to properly manage the constants. 
I have searched this article <https://semaphoreci.com/blog/constants-layer-javascript>. It provides a good manner to manage the constants in JS. 
How do you organize the constants?
Currently my idea is to create a separated `constant` package under the `internal` package and organize the constants like below:

```bash
internal
|--- constant
    |--- api
        |--- api.go
    |--- i18n
        |--- i18n.go
    |--- ...
```

**Comment by mattrltrent** [comment]
@AndrewLawrence80: your proposed structure under internal/constant seems quite reasonable (personally I'd rename it to "constants" plural, but that's preference).

I don't know fully what scale you're talking about though, however, here are some general tips I can think of that could help:

- Group constants by domain or functionality, as you’ve done with `api` and `i18n`. Extend this to other areas like errors, config, and validation.

- Use typed constants where applicable for better type safety and clarity:
  ```go
  package api
  
  type HTTPMethod string
  
  const (
      GET  HTTPMethod = "GET"
      POST HTTPMethod = "POST"
  )
  ```
- Export only the constants required outside the package and keep others un-exported to avoid unnecessary tight coupling:
  ```go
  const MaxRetries = 3 // exported (obviously because it starts with a capital)
  const defaultTimeout = 10 // non-exported 
  ```

- For constants that may change based on the environment (ie: API URLs, credentials, etc.), consider using a configuration management library. [Viper](https://pkg.go.dev/github.com/spf13/viper#section-readme) is one I like. This can help load vars/"consts" from config files. This allows you to manage such values dynamically at runtime quite cleanly.

- To state the obvious, just include lots of comments/[doc comments](https://tip.golang.org/doc/comment) explaining your constants.

- Ensure standardization among naming for consistency, like `ALL_CAPS_WITH_UNDERSCORES` for public constants and `camelCase` for private ones (or whatever your preference is).

Maybe this helps?
